### PR TITLE
Add support for special characters

### PR DESCRIPTION
**Pull request by aelkazdadi**
Support Alt mappings, as well as Ctrl+Space. getchar returns a string of the input in that case, instead of a character number, so nr2char doesn't work for them.
My solution doesn't support all possible inputs (for example inputs with accents <M-à>, etc), but this should be good enough for most people.

### CODE REVIEW

File: autoload/which_key/util.vim [hunk 1]
```diff
@@ -78,3 +78,26 @@ function! which_key#util#format(mapping) abort
...
+  return [m_char, m_char_code]
+endfunction
+
+function! s:to_list(str)
+  return split(a:str, '\zs')
+endfunction
+
+
+let s:chars = s:to_list('AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz0123456789')
+      \ + s:to_list('`~!@#$%^&*()-_=+')+ s:to_list(' 	') + s:to_list('[]{};:''"\|,./<>?')
```
Comment (liuchengxu) [inline autoload/which_key/util.vim]: To my knowledge, this is actually to get the [ASCII printable
characters](https://theasciicode.com.ar/), which could be simplified as `map(range(32, 126), 'nr2char(v:val)')`.

File: autoload/which_key.vim [hunk 2]
```diff
@@ -225,14 +225,23 @@ function! s:getchar() abort
     return ''
   endif
 
-  " <Tab>, <C-I> = 9
-  let input .= c == 9 ? '<Tab>' : nr2char(c)
+  if type(c) == s:TYPE.number
```
Comment (liuchengxu) [inline autoload/which_key.vim]: Can you extract it as a function?

Comment (aelkazdadi) [reply autoload/which_key.vim]: Should be better, now

**Comment by liuchengxu** [comment]
This PR is really a progress for which-key to take care of these special characters which is an issue that I always want to fix yet not :(. Thank you! @aelkazdadi   And #52 is related. Can you also try this patch, does it work for you? @leoatchina @Freed-Wu

**Comment by liuchengxu** [comment]
@aelkazdadi  Can you show me a minimal vimrc to try? I tried in my machine, the returns of `getchar()` of `<C-A>` and `<M-A>` are both number, 1 for `<C-A>`, 229 for `<M-A>`. I'm not sure if I'm missing something.

**Comment by aelkazdadi** [comment]
getchar doesn't return a number for <M-A> on my machine (using neovim 0.5.0-dev). Here's a minimal vimrc
```
set nocompatible
syntax on
filetype off
filetype plugin indent on
```
Result
[asciicast](https://asciinema.org/a/zROdinE7d0esJs2jkeaR1zfjc)

**Comment by liuchengxu** [comment]
Good job! 🍻

**Comment by leoatchina** [comment]
I tested the latest commit with  the following vimrc, and wrote the error in the comments.

I donot mean these errors is the defeats of vim-which-key, they may be cause by the limits of vim 

the `C-`  line which is not commented meaned it worked 
```
set nocompatible

set rtp=$HOME/.vim-plug,$VIMRUNTIME
let $PLUG_PATH = expand('$HOME/.vim/plug')
set rtp+=$PLUG_PATH
call plug#begin($PLUG_PATH)
        Plug 'liuchengxu/vim-which-key'
        set timeoutlen=500
        nnoremap <C-a>  :<C-u>WhichKey '<C-a>'<Cr>
        " C-b -> E20: mark not set
        " nnoremap <C-b>  :<C-u>WhichKey '<C-b>'<Cr>
        " C-c as cancel, not map
        " nnoremap <C-c>  :WhichKey '<C-c>'<Cr>
        nnoremap <C-d>  :<C-u>WhichKey '<C-d>'<Cr>
        nnoremap <C-e>  :<C-u>WhichKey '<C-e>'<Cr>
        " C-f -> E78: Unkown set
        " nnoremap <C-f>  :<C-u>WhichKey '<C-f>'<Cr>
        nnoremap <C-g>  :<C-u>WhichKey '<C-g>'<Cr>
        " C-h as delete, not map
        " nnoremap <C-h>  :WhichKey '<C-h>'<Cr>
        nnoremap <C-i>  :<C-u>WhichKey '<C-i>'<Cr>
        " C-j -> E115: Mssing quote
        " nnoremap <C-j>  :WhichKey '<C-j>'<Cr>
        " C-k gave ^M in command line
        " nnoremap <C-k>  :WhichKey '<C-k>'<Cr>
        nnoremap <C-l>  :<C-u>WhichKey '<C-l>'<Cr>
        " C-m -> E115 Missing quote
        " nnoremap <C-m>  :<C-u>WhichKey '<C-m>'<Cr>
        " C-n gave Whichkey' in command line
        " nnoremap <C-n>  :<C-u>WhichKey '<C-n>'<Cr>
        nnoremap <C-o>  :<C-u>WhichKey '<C-o>'<Cr>
        " C-p -> E488
        " nnoremap <C-p>  :<C-u>WhichKey '<C-p>'<Cr>
        " C-q, not tested
        "nnoremap <C-q>  :<C-u>WhichKey '<C-q>'<Cr>
        " C-r -> E115: Mssing quote
        " nnoremap <C-r>  :<C-u>WhichKey '<C-r>'<Cr>
        " C-s, not tested
        "nnoremap <C-s>  :<C-u>WhichKey '<C-s>'<Cr>
        nnoremap <C-t>  :<C-u>WhichKey '<C-t>'<Cr>
        " C-u, give ' in command line
        " nnoremap <C-u>  :<C-u>WhichKey '<C-u>'<Cr>
        nnoremap <C-v>  :<C-u>WhichKey '<C-v>'<Cr>
        " C-w -> E115 Missing quote
        " nnoremap <C-w>  :<C-u>WhichKey '<C-w>'<Cr>
        nnoremap <C-x>  :<C-u>WhichKey '<C-x>'<Cr>
        nnoremap <C-y>  :<C-u>WhichKey '<C-y>'<Cr>
        " C-z, not tested
        "nnoremap <C-z>  :<C-u>WhichKey '<C-z>'<Cr>
call plug#end()
```

**Comment by liuchengxu** [comment]
I don't take a closer look, but the vimrc you posted misused vim-which-key, you have to explicitly define the a gloabl Dict for some prefix or that prefix has some user-defined mappings already, for vim-which-key retrives the mappings's info via `:map` or the pre-defined Dict. You should not use `<C-*>` without any mapping or configuring. @leoatchina

**Comment by Freed-Wu** [comment]
> I tested the latest commit with the following vimrc, and wrote the error in the comments.
> 
> I donot mean these errors is the defeats of vim-which-key, they may be cause by the limits of vim
> 
> the `C-` line which is not commented meaned it worked
> 
> ```
> set nocompatible
> 
> set rtp=$HOME/.vim-plug,$VIMRUNTIME
> let $PLUG_PATH = expand('$HOME/.vim/plug')
> set rtp+=$PLUG_PATH
> call plug#begin($PLUG_PATH)
>         Plug 'liuchengxu/vim-which-key'
>         set timeoutlen=500
>         nnoremap <C-a>  :<C-u>WhichKey '<C-a>'<Cr>
>         " C-b -> E20: mark not set
>         " nnoremap <C-b>  :<C-u>WhichKey '<C-b>'<Cr>
>         " C-c as cancel, not map
>         " nnoremap <C-c>  :WhichKey '<C-c>'<Cr>
>         nnoremap <C-d>  :<C-u>WhichKey '<C-d>'<Cr>
>         nnoremap <C-e>  :<C-u>WhichKey '<C-e>'<Cr>
>         " C-f -> E78: Unkown set
>         " nnoremap <C-f>  :<C-u>WhichKey '<C-f>'<Cr>
>         nnoremap <C-g>  :<C-u>WhichKey '<C-g>'<Cr>
>         " C-h as delete, not map
>         " nnoremap <C-h>  :WhichKey '<C-h>'<Cr>
>         nnoremap <C-i>  :<C-u>WhichKey '<C-i>'<Cr>
>         " C-j -> E115: Mssing quote
>         " nnoremap <C-j>  :WhichKey '<C-j>'<Cr>
>         " C-k gave ^M in command line
>         " nnoremap <C-k>  :WhichKey '<C-k>'<Cr>
>         nnoremap <C-l>  :<C-u>WhichKey '<C-l>'<Cr>
>         " C-m -> E115 Missing quote
>         " nnoremap <C-m>  :<C-u>WhichKey '<C-m>'<Cr>
>         " C-n gave Whichkey' in command line
>         " nnoremap <C-n>  :<C-u>WhichKey '<C-n>'<Cr>
>         nnoremap <C-o>  :<C-u>WhichKey '<C-o>'<Cr>
>         " C-p -> E488
>         " nnoremap <C-p>  :<C-u>WhichKey '<C-p>'<Cr>
>         " C-q, not tested
>         "nnoremap <C-q>  :<C-u>WhichKey '<C-q>'<Cr>
>         " C-r -> E115: Mssing quote
>         " nnoremap <C-r>  :<C-u>WhichKey '<C-r>'<Cr>
>         " C-s, not tested
>         "nnoremap <C-s>  :<C-u>WhichKey '<C-s>'<Cr>
>         nnoremap <C-t>  :<C-u>WhichKey '<C-t>'<Cr>
>         " C-u, give ' in command line
>         " nnoremap <C-u>  :<C-u>WhichKey '<C-u>'<Cr>
>         nnoremap <C-v>  :<C-u>WhichKey '<C-v>'<Cr>
>         " C-w -> E115 Missing quote
>         " nnoremap <C-w>  :<C-u>WhichKey '<C-w>'<Cr>
>         nnoremap <C-x>  :<C-u>WhichKey '<C-x>'<Cr>
>         nnoremap <C-y>  :<C-u>WhichKey '<C-y>'<Cr>
>         " C-z, not tested
>         "nnoremap <C-z>  :<C-u>WhichKey '<C-z>'<Cr>
> call plug#end()
> ```

You should use ` nnoremap <C-w>  :<C-u>WhichKey '<lt>C-w>'<Cr>`. You can `help <lt>` to see more.

**Comment by Freed-Wu** [comment]
`<Splash>`,`<Bar>`,`<BS>`,`<Tab>` is same.  You cannot use `\` , `|` ,`^H`,`^I`.  But `"` and `<Esc>`  will return an error. And `<lt>c-<splash>` will not work. Perhaps they are not allowed?

**Comment by leoatchina** [comment]
> I don't take a closer look, but the vimrc you posted misused vim-which-key, you have to explicitly define the a gloabl Dict for some prefix or that prefix has some user-defined mappings already, for vim-which-key retrives the mappings's info via `:map` or the pre-defined Dict. You should not use `<C-*>` without any mapping or configuring. @leoatchina

Yes I know that , but the full config is quite long so just test the minimal one , @Freed-Wu have pointed out my mistakes also,  I will try to update my config.
BTW, it is no related with #79

**Comment by Freed-Wu** [comment]
I am great that my comment can help you. And I hope this plugin can support more keys. such as `"`,etc.
p.s. 
1. the `<c-tab>`  also cannot work……and press it will return an error when which_key menu has popped up. 
2. Some hotkeys like `[<space>`(be maped in unimpared.vim) and `<space><tab><tab>` cannot work…… 
3. And something I am curious that when whick_key menu poped up, I cannot press `n*`(* is any char) fast. This means I must press `<leader>nn` or `<leader> `(pause)`n`(pause)`n` not `<leader>`(pause)`nn`. The first and second can work but the third cannot work. The `<leader>` can be replaced by any keys to reappear this phenomenon.  Has any one encountered the same problem as me?

**Comment by liuchengxu** [comment]
As I have said before, the special keys are not well handled clearly in vim-which-key, for I personally do not rely on these speciall keys. I basically won't spare much time for diving into these peculiar things myself. Anyway PR is always welcome.

**Pull request merged**